server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -292,6 +292,86 @@ public ShardIterator activeInitializingShardsRankedIt(
+     */
+    public ShardIterator activeInitializingShardsWeightedIt(
+        WeightedRouting weightedRouting,
+        DiscoveryNodes nodes,
+        WeightedRoutingCache cache,
+        int defaultWeight
+    ) {
+        final int seed = shuffler.nextSeed();
+        List<ShardRouting> ordered = new ArrayList<>();
+        List<ShardRouting> orderedActiveShards = cache.get(new WeightedRoutingCache.Key(shardId));
+        if (orderedActiveShards == null) {
+            orderedActiveShards = shardsOrderedByWeight(activeShards, weightedRouting, nodes, defaultWeight);
+            cache.put(new WeightedRoutingCache.Key(shardId), orderedActiveShards);
+        }
+        ordered.addAll(shuffler.shuffle(orderedActiveShards, seed));
+
+        if (!allInitializingShards.isEmpty()) {
+            List<ShardRouting> orderedInitializingShards = shardsOrderedByWeight(
+                allInitializingShards,
+                weightedRouting,
+                nodes,
+                defaultWeight
+            );
+            ordered.addAll(orderedInitializingShards);
+        }
+        return new PlainShardIterator(shardId, ordered);
+    }
+
+    /**
+     * Returns a list containing shard routings ordered using weighted round-robin scheduling.
+     */
+    private List<ShardRouting> shardsOrderedByWeight(
+        List<ShardRouting> shards,
+        WeightedRouting weightedRouting,
+        DiscoveryNodes nodes,
+        int defaultWeight
+    ) {
+        WeightedRoundRobin<ShardRouting> weightedRoundRobin = new WeightedRoundRobin<>(
+            calculateShardWeight(shards, weightedRouting, nodes, defaultWeight)
+        );
+        List<WeightedRoundRobin.Entity<ShardRouting>> shardsOrderedbyWeight = weightedRoundRobin.orderEntities();
+        List<ShardRouting> orderedShardRouting = new ArrayList<>(activeShards.size());
+        if (shardsOrderedbyWeight != null) {
+            for (WeightedRoundRobin.Entity<ShardRouting> shardRouting : shardsOrderedbyWeight) {
+                orderedShardRouting.add(shardRouting.getTarget());
+            }
+        }
+        return orderedShardRouting;
+    }
+
+    /**
+     * Returns a list containing shard routing and associated weight. This function iterates through all the shards and
+     * uses weighted routing to find weight for the corresponding shard. This is fed to weighted round-robin scheduling
+     * to order shards by weight.
+     */
+    private List<WeightedRoundRobin.Entity<ShardRouting>> calculateShardWeight(
+        List<ShardRouting> shards,
+        WeightedRouting weightedRouting,
+        DiscoveryNodes nodes,
+        int defaultWeight
+    ) {
+        List<WeightedRoundRobin.Entity<ShardRouting>> shardsWithWeights = new ArrayList<>();
+        for (ShardRouting shard : shards) {
+            DiscoveryNode node = nodes.get(shard.currentNodeId());
+            String attVal = node.getAttributes().get(weightedRouting.attributeName());
```

Comment:
Can we cache node's attr val, this might turn out to an optimisation for lesser nodes stuffed with 1000s of shards